### PR TITLE
[NC-2138] Implement chain download for fast sync

### PR DESCRIPTION
**Pull request by ajsutton**
## PR description
Implements a `FastSyncChainDownloader` and hooks it into `FastSyncActions` so that the headers and bodies for the chain up to the pivot point are downloaded.  It deliberately avoids downloading blocks that are beyond the pivot block.

Receipts aren't yet downloaded and we're still performing full validation on all headers rather than light validation.

### CODE REVIEW

File: ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/CheckpointHeaderManager.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+              return getCheckpointsAwaitingImport();
+            });
+  }
+
+  protected CompletableFuture<List<BlockHeader>> getAdditionalCheckpointHeaders(
+      final SyncTarget syncTarget, final BlockHeader lastHeader) {
+    return requestAdditionalCheckpointHeaders(lastHeader, syncTarget)
+        .handle(
+            (headers, t) -> {
+              t = ExceptionUtils.rootCause(t);
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/CheckpointHeaderManager.java]: Of course our throwable accepts a null param and returns null in that case...

[the remark this answers is not in the document]
Comment (ajsutton) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/CheckpointHeaderManager.java]: Seems like a good idea to delegate to Guava's implementation but the null handling is really very useful given the API that `CompletableFuture` uses often gives us `null` exceptions to deal with: https://github.com/PegaSysEng/pantheon/pull/692

[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/CheckpointHeaderManager.java]: beat me to it - https://github.com/PegaSysEng/pantheon/pull/693

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncCheckpointHeaderManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        checkpointHeaderManager.pullCheckpointHeaders(syncTarget);
+
+    final Responder responder = RespondingEthPeer.blockchainResponder(otherBlockchain);
+    peer.respondWhile(responder, peer::hasOutstandingRequests);
+
+    assertThat(result).isCompletedWithValue(asList(block(10), block(15), pivotBlockHeader));
+  }
+
+  @Test
+  public void shouldNotDuplicatePivotBlockAsCheckpoint() {
```
Comment (mbaxter) [inline ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncCheckpointHeaderManagerTest.java]: Might be worth adding a few more edge cases here - when the common ancestor is at the block prior to the pivot, or at the pivot etc.

File: ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncTargetManager.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        return completedFuture(Optional.empty());
+      } else {
+        return confirmPivotBlockHeader(bestPeer);
+      }
+    }
+  }
+
+  private CompletableFuture<Optional<EthPeer>> confirmPivotBlockHeader(final EthPeer bestPeer) {
+    final RetryingGetHeaderFromPeerByNumberTask task =
+        RetryingGetHeaderFromPeerByNumberTask.forPivotBlock(
```
[the remark this answers is not in the document]
Comment (ajsutton) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncTargetManager.java]: Done.

File: ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncTargetManager.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            pivotBlockHeader.getNumber(),
+            MAX_PIVOT_BLOCK_RETRIES);
+    task.assignPeer(bestPeer);
+    return ethContext
+        .getScheduler()
+        .timeout(task)
+        .thenApply(
+            result -> {
+              if (result.size() != 1 || !result.get(0).equals(pivotBlockHeader)) {
+                bestPeer.disconnect(DisconnectReason.USELESS_PEER);
```
[the remark this answers is not in the document]
Comment (ajsutton) [reply ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncTargetManager.java]: Yeah I went back and forward on it.  If we don't disconnect we'll need to maintain a sync target blacklist to avoid continuously selecting the same peer on a different chain.  But ultimately I'm ok with disconnecting because while we might be useful data from a peer it's seems like a pretty strong indicator that something is wrong with a peer that's forked 500 blocks back in disagreement with the majority of our other peers.

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncCheckpointHeaderManagerTest.java (hunk 2 above)
Comment (ajsutton) [reply ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncCheckpointHeaderManagerTest.java]: Done.
